File: nbody.py
"""N-body problem, implemented in pure Python.

This is just for demo purposes; any sensible implementation should
be based on Cython, Numba, or a GPGPU library.
"""
from __future__ import annotations
from dataclasses import dataclass
import math
import random
from typing import List


@dataclass
class Vector:
    """Generic 3D vector
    """

    __slots__ = ("x", "y", "z")

    x: float
    y: float
    z: float

    def __add__(self, other: Vector) -> Vector:
        return Vector(self.x + other.x, self.y + other.y, self.z + other.z)

    def __sub__(self, other: Vector) -> Vector:
        return Vector(self.x - other.x, self.y - other.y, self.z - other.z)

    def __iadd__(self, other: Vector) -> Vector:
        self.x += other.x
        self.y += other.y
        self.z += other.z
        return self

    def __isub__(self, other: Vector) -> Vector:
        self.x -= other.x
        self.y -= other.y
        self.z -= other.z
        return self

    def __mul__(self, other: float) -> Vector:
        return Vector(self.x * other, self.y * other, self.z * other)

    def __truediv__(self, other: float) -> Vector:
        return Vector(self.x / other, self.y / other, self.z / other)

    @property
    def lensq(self) -> float:
        return self.x ** 2 + self.y ** 2 + self.z ** 2


@dataclass
class Body:
    """State of a celestial object at a certain moment in time
    """

    __slots__ = ("mass", "position", "velocity")

    mass: float
    position: Vector
    velocity: Vector

    def attraction(self, other: Body) -> Vector:
        """Calculate attraction force between two bodies
        """
        dist = self.position - other.position
        dist_modsq = dist.lensq
        dist_unit = dist / math.sqrt(dist_modsq)  # Unit vector
        G = 6.674384e-11
        force_mod = G * self.mass * other.mass / dist_modsq
        return dist_unit * force_mod

    def move(self, force: Vector, dt: float) -> Body:
        """Apply force for dt seconds to this body.
        Return new body with updated velocity and position.
        This object is left unchanged.
        """
        velocity = self.velocity + force / self.mass * dt
        position = self.position + velocity * dt
        return Body(self.mass, position, velocity)


def nbody(bodies: List[Body], dt: float) -> List[Body]:
    """Have N bodies interact with each other, each applying gravitational force to
    each other for dt seconds and thus changing each other's position and velocity.

    :param bodies:
        Initial state of the bodies
    :param float dt:
        Time increment (in seconds). It should be set small enough that any change in
        velocity and position occurring during it should not materially impact the
        next iteration.
    :returns:
        New state of the bodies after dt. The initial state is left unaltered.
    """
    forces = [Vector(0.0, 0.0, 0.0) for _ in bodies]
    for b1, f1 in zip(bodies, forces):
        for b2, f2 in zip(bodies, forces):
            if b1 is not b2:
                force = b1.attraction(b2)
                f1 += force
                f2 -= force
    return [body.move(force, dt) for body, force in zip(bodies, forces)]


def rand_bodies(n: int) -> List[Body]:
    """Generate n bodies at rest, with random uniform weight distribution up to
    1 metric tonne, uniformly distributed within a 1 cubic metre space
    """
    return [
        Body(
            mass=random.uniform(0, 1000),
            position=Vector(random.random(), random.random(), random.random()),
            velocity=Vector(0.0, 0.0, 0.0),
        )
        for _ in range(n)
    ]


if __name__ == "__main__":
    bodies = rand_bodies(2)
    print(bodies)
    for _ in range(5):
        bodies = nbody(bodies, 0.1)
        print(bodies)

File: nbody.cpp
#include <cmath>
#include <vector>

const double G = 6.674384e-11;


class Vector3 {
public:
    double x, y, z;

    Vector3(): x(0.0), y(0.0), z(0.0) {};
    Vector3(double x, double y, double z): x(x), y(y), z(z) {};
    Vector3(const Vector3 & other): x(other.x), y(other.y), z(other.z) {};

    void clear() {
        x = 0.0;
        y = 0.0;
        z = 0.0;
    };

    Vector3 operator+(const Vector3 & other) const {
        return Vector3(x + other.x, y + other.y, z + other.z);
    };

    Vector3 operator-(const Vector3 & other) const {
        return Vector3(x - other.x, y - other.y, z - other.z);
    };

    Vector3& operator+=(const Vector3 & other) {
        x += other.x;
        y += other.y;
        z += other.z;
        return *this;
    };

    Vector3& operator-=(const Vector3 & other) {
        x -= other.x;
        y -= other.y;
        z -= other.z;
        return *this;
    };

    Vector3 operator*(const double other) const {
        return Vector3(x * other, y * other, z * other);
    };

    Vector3 operator/(const double other) const {
        return Vector3(x / other, y / other, z / other);
    };

    Vector3& operator*=(const double other) {
        x *= other;
        y *= other;
        z *= other;
        return *this;
    };

    Vector3& operator/=(const double other) {
        x /= other;
        y /= other;
        z /= other;
        return *this;
    };
};


class Body {
public:
    double mass;
    Vector3 position;
    Vector3 velocity;

    Body(): mass(0.0), position(), velocity() {};

    Body(double mass, Vector3& position, Vector3& velocity):
        mass(mass), position(position), velocity(velocity) {};
};


class NBody {
public:
    NBody() {};
    void move(double dt);

    std::vector<Body> bodies;
private:
    std::vector<Vector3> forces;
};


void NBody::move(double dt) {
    auto n = bodies.size();

    if (forces.size() != n) {
        forces.resize(n);
    }
    for (auto it = forces.begin(); it != forces.end(); it++) {
        it->clear();
    }

    for (auto i=0; i < n; i++) {
        for (auto j=0; j < n; j++) {
            if (i != j) {
                /*
                   dist vector = position vector[i] - position vector[j]

                                       dist vector
                   dist unit vector = -------------
                                      |dist vector|

                                 mass[i] * mass[j]
                   |force| = G * -----------------
                                  |dist vector|^2

                   force vector = |force| * dist unit vector
                */

                auto dist_v = bodies[i].position - bodies[j].position;
                auto dist_modsq = dist_v.x * dist_v.x + dist_v.y * dist_v.y + dist_v.z * dist_v.z;
                auto dist_mod = sqrt(dist_modsq);
                auto force_mod = G * bodies[i].mass * bodies[j].mass / dist_modsq;
                auto force_v = dist_v / (dist_mod / force_mod);
                forces[i] += force_v;
                forces[j] -= force_v;
            }
        }
    }

    for (size_t i=0; i < n; i++) {
        auto& bi = bodies[i];
        auto dv = forces[i] / (bi.mass / dt);
        auto ds = (bi.velocity + dv / 2.0) * dt;
        bi.position += ds;
        bi.velocity += dv;
    }
}
